connectors/sources/sharepoint_online.py

```diff
@@ -605,22 +605,83 @@ async def site_users(self, site_web_url):
         except NotFound:
             return
 
-    async def sites(self, parent_site_id, allowed_root_sites):
-        select = ""
+    async def sites(
+        self,
+        parent_site_id,
+        allowed_root_sites,
+        enumerate_all_sites=True,
+        fetch_subsites=False,
+        subsite_depth=3,
+    ):
+        if allowed_root_sites == [WILDCARD] or enumerate_all_sites:
+            self._logger.debug(f"Looking up all sites to fetch: {allowed_root_sites}")
+            async for site in self._all_sites(parent_site_id, allowed_root_sites):
+                yield site
+        else:
+            self._logger.debug(f"Looking up individual sites: {allowed_root_sites}")
+            expand_clause = self._subsite_expansion_str(subsite_depth)
+            for allowed_site in allowed_root_sites:
+                try:
+                    if fetch_subsites:
+                        async for site in self._fetch_site_and_subsites(
+                            parent_site_id, allowed_site, expand_clause
+                        ):
+                            yield site
+                    else:
+                        yield await self._fetch_site(parent_site_id, allowed_site)
+
+                except NotFound:
+                    self._logger.warning(
+                        f"Could not look up site '{allowed_site}' by relative path in parent site: {parent_site_id}"
+                    )
 
+    async def _all_sites(self, parent_site_id, allowed_root_sites):
+        select = ""
         async for page in self._graph_api_client.scroll(
             f"{GRAPH_API_URL}/sites/{parent_site_id}/sites?search=*&$select={select}"
         ):
             for site in page:
                 # Filter out site collections that are not needed
-                if (
-                    WILDCARD not in allowed_root_sites
-                    and site["name"] not in allowed_root_sites
-                ):
+                if [WILDCARD] != allowed_root_sites and site[
+                    "name"
+                ] not in allowed_root_sites:
                     continue
-
                 yield site
 
+    async def _fetch_site_and_subsites(
+        self, parent_site_id, allowed_site, expand_clause
+    ):
+        self._logger.debug(
+            f"Requesting site '{allowed_site}' and subsites by relative path in parent site: {parent_site_id}"
+        )
+        site_with_subsites = await self._graph_api_client.fetch(
+            f"{GRAPH_API_URL}/sites/{parent_site_id}:/sites/{allowed_site}?{expand_clause}"
```

Comment:
WDYT if we [enumerate subsites](https://learn.microsoft.com/en-us/graph/api/site-list-subsites) separately? This will definitely increase the number of requests, but it does have some benefits:
1. Easier to understand
2. Less configuration (no need to config the depth)
3. Confirm all subsites are fetched, assuming there's limit in expand result.

Reply:
Ok, I've implemented a recursion like this that doesn't limit depth, but also doesn't make useless requests at leaf nodes. WDYT?

Reply:
I think we're in a good state of compromise for now. It can always be improved in a future effort.